Add ProjectileEffect base class and basic functionality

Checkroth opened the pull request:
Adds a new Node type and class called ProjectileEffect. Includes the usage of this class in the player and projectile scenes.

Will also include the following projectile effects:

- Double laser
- Big Laser
- Fast Laser

Checkroth committed:
Add ProjectileEffect base class and basic functionality

Adds a new Node type and class called ProjectileEffect.
Includes the usage of this class in the player and projectile scenes.

Does not include any actual projectile effects yet.

campenr commented on units/player.gd:
```diff
@@ -32,6 +34,7 @@ func _physics_process(delta):
-		owner.add_child(p)
+		p.spawn(owner, projectile_effects)
+		# TODO:: This transform should probably be modifiable by projectile effects too?
 		p.transform = projectile_spawner.global_transform
```
All this is doing currently is making sure the projectile doesn't move with the player as the player moves.
I.e. it moves relative to the global space, not relative to the player.

Definitely can do other changes to this transform afterwards with effects 👍

campenr commented on units/projectile.gd:
```diff
@@ -1,6 +1,22 @@
 func _physics_process(delta):
-	position += transform.x * SPEED * delta
+	var position_modifier = transform.x * SPEED * delta
+	for effect in projectile_effects:
```
Will all effects get applied every tick? Or will some only apply every x-seconds / ticks? Thinking i.e. a projectile that splits periodically. Or explodes after a certain amount of time.
Or would we have a "timer" effect that chains with something that will happen after that time? 🤔

Checkroth replied:
I think whether or not the effect needs to o something should be a responsibility of the effect.

I've since pushed a few changes that basically involve just passing everything relate to the operation down the callstack, which is kinda ugly but it does let us obfuscate the underlying details

campenr reviewed (approved):
🚢  but with some random questions that occured to me while reading the PR

Checkroth committed:
Add a double laser projectile effect.

This effect adds an identical lazer ouput offset from the original
laser.
It also makes some modifications to how the firing mechanism and
effects dispatch is handled, to make it easier to offload effect
mechanism management to each effect.

Checkroth commented on units/player.gd:
```diff
@@ -9,12 +9,27 @@ const HORIZONTAL_BUFFER = 75
+	#    utility functions will get called regardless of its presense in the scene
+	#    tree, as long as thy're in the top-level projectile_effects Array.
+	add_child(double_laser.instantiate())
+	projectile_effects.append(double_laser)
```
Note:

This is not actually appending anything to the top-level projectile_effects array. Later instances where this is accessed see an empty array for some reason :shrug:

Checkroth committed:
Add Huge Laser projectile effect

This adds a projectile effect that makes lasers huge.
It includes refactoring to the projectile effect engine to allow
effects to directly influence the parent node's size

Checkroth committed:
Add projectile scaling workaround

Adds a workaround that manually scales the children of the Projectile
scene.
This is due to a problem where the projectile root node scale does not
propagate down to the sprite or collision shape within it.

https://github.com/godotengine/godot/issues/5734

Checkroth committed:
Refactor to use packedscene instead of node type

This refactors the projectile effect system to use packed scenes
instead of the base class node type ProjectileEffect.

This is due to the requirement for these scenes to be instantiated for
each projectile, not a shared resource. Without this change, effects
that modify the scene tree (such as HUGE LASER) do not work.

Checkroth committed:
Add Fast Laser Effect

Adds a new effect that hooks in to the physics system to make lasers
go pewpew faser

Checkroth commented:
Re-requesting review because the PR has evolved significantly and now is now officially "done"

Checkroth commented on projectiles/projectile_effect.gd:
```diff
@@ -0,0 +1,75 @@
+
+# Modify the logic that adds the projectile as a child to the scene.
+# This can be used to change spawn locations, quantities, etc.
+func modify_creation(owner: Node, projectile_effects: Array, transform: Transform2D):
```
:wine_glass: :cheese: 

I'm pretty sure the resulting type here is "Packed Scene", which currently [cannot be typed better](https://github.com/godotengine/godot-proposals/issues/8565)

I flip-flopped between the actual node type (ProjectileEffect) and PackedScenes and some other ideas so many times that changing every type annotation was incredibly cumbersome, so I've opted to leave this as just `Array` until we flesh things out and settle on something that works and won't be changed.

There's also no support for Type Variables -- I figured having a Globals script that contains this type so we can just say `var EffectType = Array[PackedScene]` and use that so that making a change would be trivial, but can't find a way to do it with the current gdscript version.

campenr replied:
💬 Some interesting thoughts in this forum post on accessing more info about types when using PackedScene. A custom resource type is an interesting approach. Nothing for now though, just food for collective thought.

campenr replied:
Classic me. No link, and now I can't find it again 🤦

Checkroth commented on units/player.gd:
```diff
@@ -9,11 +9,39 @@ const HORIZONTAL_BUFFER = 75
-
+	_add_test_effects()
+
+func _add_test_effects():
```
:wine_glass: :cheese: 

This, an the top-level `projectile_effects` array is not used (and therefore this function effectively does nothing)

This is for two reasons:

1. The core node type `ProjectileEffect` is not used now, see the comment about using PackedScene instead.
2. Passing _just_ the type `ProjectileEffect` (or a subclass -- HugeLaser, DoubleLaser, etc.) results in an empty array for a reason I truly have no idea about. Otherwise I would keep the array, use a non-instantiated variable, and then use that instead of the `PackedScene`

I think ultimately, a combination of ideas might be the best path forward: A `Globals` script that keeps a list of all our "officially supported" effects as an enum & dictionary of that Enum to a resource path (e.g. `res://projectiles/fast_laser.tscn`), and then maintain the top level list of just the enum, and offload the loading/instansiating to the projectile scene.

Checkroth commented on projectiles/huge_laser.gd:
```diff
@@ -0,0 +1,2 @@
+class_name HugeLaser
+extends ProjectileEffect
```
:wine_glass: :cheese: 

This is not actually needed with the current implementation.

It WILL be needed if we want to use the Node class type instead of a PackedScene for effects management. I.e. you can't do `HugeLaser.new()` unless you define the class_name in a script.
But if you load a `PackedScene` from the scene path, this is totally superfluous boilerplate.

Checkroth commented on units/projectile.gd:
```diff
@@ -1,6 +1,41 @@
+	scale = new_scale
+
+func _ready():
+	if scale != Vector2.ONE:
```
:wine_glass: :cheese: 

This is a hack to allow the "scale" you set on the Projectile node itself to impact the actual scale of the base projectile.
Without this, setting "Transform -> Scale" on the base projectile node will change the size of the projectile in the 2D visual UI, but at runtime it will be completely ignored. See [this godot pr](https://github.com/godotengine/godot/issues/5734), also linked in the commit message directly.

Checkroth commented on units/player.gd:
```diff
@@ -30,8 +58,5 @@ func _physics_process(delta):
-		var p = Projectile.instantiate()
-		owner.add_child(p)
-		p.transform = projectile_spawner.global_transform
+		fire()
```
:wine_glass: :cheese: 

Unnecessary change ultimately, but was done in an early stage of this PR implementation that played with the idea of projectile effects calling `fire` directly on the player scene. That didn't work the way I wanted it to, but I decided to keep the fire mechanism as its own function because it _feels_ right, and if we ever want to do a little more with this mechanism we'd probably have to do this refactor anyway.

campenr replied:
Eventually we'll want to add some state machine here to tidy this all up 👍 meantime this is a nice prefactor :D

campenr commented on projectiles/double_laser.gd:
```diff
@@ -0,0 +1,33 @@
+	Projectile = load("res://units/projectile.tscn")
+# Called when the node enters the scene tree for the first time.
+
+# Called every frame. 'delta' is the elapsed time since the previous frame.
```
🗑️

campenr commented on projectiles/double_laser.gd:
```diff
@@ -0,0 +1,33 @@
+class_name EffectDoubleLaser
+extends ProjectileEffect
+
+var Projectile : PackedScene = preload("res://units/projectile.tscn")
```
❓ We're preloading the scene here, but also calling `load` in `_ready`. Do we need both? 🤔

Checkroth replied:
nope

campenr commented on projectiles/double_laser.gd:
```diff
@@ -0,0 +1,33 @@
+		return not effect.is_double_laser
+
+	# TODO:: Hacky? but const on the node does not work for packed scenes
+	return effect.resource_path != "res://projectiles/double_laser.tscn"
```
❓ Non-blocking: can we type check the effect? I guess we're loading the scene, but we could probably type check the parent node of the scene to be able to check against a custom `class_name`? 🤔

campenr commented on projectiles/projectile_effect.gd:
```diff
@@ -0,0 +1,75 @@
+#    in order of Effect Priority
+# TODO:: Effect Priority as an enum? (to discourage "priority: 9999999")
+
+# Each modifier is a combination of a double value and a modification type.
```
❓ Non-blocking: Do we want to only do additive? In that it also allows multiplicative i.e. multiple additive nodes can be applied to get multiplicative effects.

Checkroth replied:
Not sure I follow.

Additive can be achieved if the projectile effect is the only one modifying that attribute of the projectile, but would not work if we wanted two different effects to compound on eachother.

Ex: We have a "speed multiplier" effect that makes the projectile go 10x faster, and we have a "rate stack effect" pickup, for each one you pick up you get +5 to fire rate and projectile speed.

So you could have 3 additive stacks, followed by a multiplier

campenr replied:
Yep fair point 👍

campenr commented on units/player.gd:
```diff
@@ -9,11 +9,39 @@ const HORIZONTAL_BUFFER = 75
+func _add_test_effects():
+		# Debugging -- these should be added to the list via some user interaction instead.
+	var double_laser = load("res://projectiles/double_laser.tscn")
+	# TODO:: Not even sure if adding to the scene tree here is necessary.
```
👍 It's necessary for lifecycle things. If we want to listen to signals etc. 👍

campenr reviewed (approved):
Looks good! I like that adding multiple of the same effect works as expected (i..e super super fast, super super big) 👍 👍 👍 

Not going to block. But do we want to consider embracing a clean commit style approach here to make it easier to see how a thing was done start -> (a) finish?.

campenr commented:
👍

On Mon, Aug 19, 2024, 5:20 PM Charles Heckroth ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In projectiles/projectile_effect.gd
> <https://github.com/campenr/radius-game/pull/2#discussion_r1721224733>:
>
> > @@ -0,0 +1,75 @@
> +class_name ProjectileEffect
> +extends Node
> +# A ProjectileEffect is a set of post-processors meant to be run on every instance of a Projectile.
> +# This by itself does nothing but is meant to be extended by various types of projectile effect.
> +# Each effect will compound on the others. Each function will be called by the Projectile scene,
> +#    in order of Effect Priority
> +# TODO:: Effect Priority as an enum? (to discourage "priority: 9999999")
> +
> +# Each modifier is a combination of a double value and a modification type.
>
> Not sure I follow.
>
> Additive can be achieved if the projectile effect is the only one
> modifying that attribute of the projectile, but would not work if we wanted
> two different effects to compound on eachother.
>
> Ex: We have a "speed multiplier" effect that makes the projectile go 10x
> faster, and we have a "rate stack effect" pickup, for each one you pick up
> you get +5 to fire rate and projectile speed.
>
> So you could have 3 additive stacks, followed by a multiplier
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/campenr/radius-game/pull/2#discussion_r1721224733>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACTFU5YAQQPTT7PPMMQECCLZSF6BFAVCNFSM6AAAAABMU65ZY2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDENBUGYZTKMBXHE>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

Checkroth commented:
> consider embracing a clean commit style approach here

I thought that's what I was doing :laughing: maybe my commits aren't pristine, but I did intend to do it atomically.

There's some refactoring in here that could be squashed in to earlier commits, but it was also a lot of discovery and small changes that were rebase conflict magnets, which I opted to just not deal with

Checkroth committed:
Move projectiles to layer 2, add enemy mask 2

This moves projectiles to a new collision layer, and adds that layer
to the mask for enemies.

This is to do two things: (1) Prevent lasers from colliding with
themselves ("Huge" and "Double" lasers can overlap), (2) allow enemies
to queue_free themselves, instead of leaving that burden on the projectile

campenr commented on units/projectile.tscn:
```diff
@@ -8,6 +8,7 @@ radius = 4.0
 
 [node name="Projectile" type="Area2D"]
 position = Vector2(68, 1)
+collision_layer = 2
```
🤔 I kinda like the idea that projectiles can hit themselves / the player... but that'll take some thinking. For now 👍 👍

Checkroth replied:
Not a terrible idea, though the huge double lasers overlapped hitboxes, so they would just immediately disappear

campenr reviewed (approved):
LGTM!!! 🚢 🚢 🚢 🚢
